### PR TITLE
ImageDreamfusion accumulate grad, guidance_eval in sd,if

### PR DESCRIPTION
**Pull request by voletiv**
1. Accumulates grad in imagedreamfusion
Fixes imagedreamfusion to accumulate grads from ref and guidance

2. Adds guidance_eval to stable_diffusion, deep_floyd
Adds the ability to optionally evaluate guidance periodically, in stable_diffusion and deep_floyd (as done in zero123)

This does not affect current working of stable_diffusion, deep_floyd, it only adds the option to visualize the guidance.

## Example

I used the following command to check the results with stable diffusion guidance:
`python launch.py --config configs/experimental/imagecondition.yaml --train --gpu 0 system.freq.ref_only_steps=20 system.freq.guidance_eval=13`

To run with deep floyd guidance, I uncommented the requisite lines in `configs/experimental/imagecondition.yaml`.

First, since `ref_only_steps=20`, it should run with just rgb_loss for 20 iterations.
Then, from 21st iteration, it should also compute guidance loss.
Then, since `freq.guidance_eval=13`, at the 26th iteration, it should save the results of guidance_eval:

With stable-diffusion:
it26-train

With deep-floyd:
it52-train

### CODE REVIEW

**Review by bennyguo — COMMENTED** [review]
I generally found that functions with a varied-length return value can be difficult to maintain. Should we consider moving `guidance_eval_out` to the return dict, for example:
```python
  guidance_out = {
      "loss_sds": loss_sds,
      "grad_norm": grad.norm(),
      "eval": the_eval_results
  }
```

File: threestudio/systems/imagedreamfusion.py [hunk 1]
```diff
@@ -109,63 +119,109 @@ def training_step(self, batch, batch_idx):
...
                     "Normal is required for normal smooth loss, no normal is found in the output."
                 )
-            normal = out["normal"]
-            loss_normal_smooth = (
-                normal[:, 1:, :, :] - normal[:, :-1, :, :]
-            ).square().mean() + (
-                normal[:, :, 1:, :] - normal[:, :, :-1, :]
-            ).square().mean()
-            self.log("train/loss_normal_smooth", loss_normal_smooth)
-            loss += self.C(self.cfg.loss.lambda_normal_smooth) * loss_normal_smooth
```
Comment (claforte) [inline threestudio/systems/imagedreamfusion.py:151]: Was this removed on purpose? I recommend keeping this loss as well as 3d_normal_smooth in case `normal_smooth` is much cheaper computationally (less rays) and still provides a good regularization.

**Review by claforte — CHANGES_REQUESTED** [review]
Thanks Vikram! Most of it looks good to me!
Please address Ben's review and restore the normal smooth (unless there's a good reason to remove it).

**Comment by claforte** [comment]
@bennyguo this introduces redundancies but I think it's a good short-term step... so unless you have a concern, we'd like to merge it today.
Later I recommend we fold some of the loss calculation/logging code in the base system class and maybe fold some of the eval guidance logic in the base guidance class, to avoid unnecessary code duplication. How does that sound?

File: threestudio/systems/imagedreamfusion.py (hunk 1 above)
Comment (voletiv) [reply threestudio/systems/imagedreamfusion.py:151]: No it's present in the lines above 160-165. Just a mismatch in tracking changes

**Comment by voletiv** [comment]
I've folded guidance_eval_save into base.py, and address the comments. Please check!

**Review by claforte — APPROVED** [review]
LGTM! Just test that DeepFloyd still works, as we discussed. :-)

**Comment by voletiv** [comment]
Tested!
Merging with main
